### PR TITLE
Issue #4156 Remove use of PlaceHolderSession for simultaneous session loading

### PR DESCRIPTION
**Pull request by janbartel**
#4156

The PlaceHolderSession was used to control for simultaneous threads attempting to load the same session, and ensuring that only one would succeed on the load, and all others would use that same session.  I've replaced that code with the use of the ConcurrentHashMap.computeIfAbsent semantic:  only one thread should succeed in loading the session, and all other threads should use that same session.

### CODE REVIEW

**Comment by sbordet** [comment]
I'm also unclear at why we need this PR, whether it's code cleanup or an actual fix.

If the latter, I don't understand what the problem was - I read the issue comments, but none refers to the session placeholder as being the problem and hence needed to be removed?

File: jetty-server/src/main/java/org/eclipse/jetty/server/session/AbstractSessionCache.java [hunk 1]
```diff
@@ -343,113 +341,60 @@ public Session get(String id) throws Exception
     protected Session getAndEnter(String id, boolean enter) throws Exception
     {
         Session session = null;
-        Exception ex = null;
 
-        while (true)
-        {
-            session = doGet(id);
+        session = doGet(id);
```
Comment (gregw) [inline jetty-server/src/main/java/org/eclipse/jetty/server/session/AbstractSessionCache.java]: You should just do computeIfAbsent here as it has an implicit doGet and act if null already

File: jetty-server/src/main/java/org/eclipse/jetty/server/session/AbstractSessionCache.java [hunk 2]
```diff
@@ -343,113 +341,60 @@ public Session get(String id) throws Exception
...
-                    //got the session
+                    session.setResident(true); //ensure freshly loaded session is resident
                     if (enter)
                         session.use();
-                    break;
+                }
+            }
+        }
+        else
+        {
```
Comment (gregw) [inline jetty-server/src/main/java/org/eclipse/jetty/server/session/AbstractSessionCache.java]: You wont need this else branch if you just go straight to computeIfAbsent above.

**Comment by janbartel** [comment]
The motivation for this PR is the problem reported in #4156. The analysis shows that 2 simultaneous requests entered the same context both with the same session id. The logs show that session was previously correctly scavenged after timing out. Thread 1 correctly finds that the session does not exist in either the cache or the backing store and then wants to create a new session. When creating a new session, the id present on the request will be used for the new session, iff that id is in use by at least one context. The test for that is if the cache/backing store for any SessionHandler reports that it has a key with that id. Just before thread 1 asks the question, thread 2 comes along and tries to load the (non-existant) session from the backing store. This load operation uses the PlaceHolder semantic to ensure that only 1 thread will succeed in doing the load, and all other threads will be blocked until the load completes and subsequently use that same session object.  The problem ultimately is that thread 2's PlaceHolder is seen by thread 1's call to isSessionInUse() and thus reports a false positive that the session exists.

This change is to remove the use of the complicated PlaceHolder code and replace it by the jvm's ConcurrentHashMap.computeIfAbsent() method, whose behaviour guarantees that 1 and only 1 thread will ever perform the computation (ie loading the session) and all other threads will use the results of the computation (ie the loaded session object).

**Comment by janbartel** [comment]
Offline there was some discussion about whether the AbstractSessionCache.exists() method needs to use the new computeIfAbsent() semantic to ensure that we don't get any "false" negatives: ie a "get" call on the session map returns null because computeIfAbsent() has not finished yet.

I contend that it does _not_ and that "false" negatives are not a problem. The computeIfAbsent() method is useful for ensuring only 1 thread calls load on the session data and that data is then used by all callers. In the case of the exists() method, we don't care _at all_ about the session data: we just want to know if the id exists && isvalid && !expired.

If you look at the code, we first do a get() on the session map:  iff that returns null (ie session is not in use for this context on this node), then we go and ask the backing store if the session exists. Thus, even if some other thread is in the process of doing a computeIfAbsent() to load the session into the cache but not yet finished, and we got a "false" negative, we would still detect the existence of the session because we ask the store.

Note also that the question we ask the store does _not_ necessarily entail a full load of the session data: for those stores that can do more fine-grained data-reads, such as the jdbc store, it can just check if the key exists, the isvalid and expiry fields in the table.  We would not want to do a full deserialization of all the session data just to ask the question does the id exist.

File: jetty-server/src/main/java/org/eclipse/jetty/server/session/AbstractSessionCache.java (hunk 2 above)
Comment (janbartel) [reply jetty-server/src/main/java/org/eclipse/jetty/server/session/AbstractSessionCache.java]: Done.

**Pull request merged**